### DISCUSSION TITLE
Smart contracts and App development

### DISCUSSION THREAD

**Discussion by paOol** [ava-labs/hypersdk#89] [Q&A] [answered]
What does the process look like when you get higher up on the stack?

What kind of tooling/SDKs would work with a hyperVM , what kind wouldn't?

Lets say you wanted a developer experience similar to solidity, but uses WASM instead. What would be required on the VM side to support this? Would a smart contract written in GO for example, be deployable on every hypervm (assuming wasm compatibilty is built in)? 

Or are HyperVMs intentionally meant for  RPC calls only?

**Comment by TacitvsXI** [comment, marked as answer]
Hey, @paOol 

If you're using a HyperVM, it's important to make sure that any software or SDKs you're using are compatible with the virtualization technology you're using. You can check the documentation or contact the vendor to make sure. Some popular SDKs that work well with HyperVM include Docker SDK, Kubernetes SDK, and OpenStack SDK.

If you want to create a smart contract experience like Solidity but using WASM, you'll need to create a virtual machine (VM) that can execute WASM bytecode and provide the necessary interfaces for deploying and interacting with smart contracts. This could involve building your own runtime environment or using an existing VM like wasmtime or wasmer. Depending on your specific needs, you might need to add additional features or capabilities to your VM.

Overall, HyperVMs can be used for many different applications beyond just remote procedure calls (RPCs). But it's important to choose the right tools and SDKs to ensure that everything works smoothly and efficiently.

**Comment by patrick-ogrady** [comment]
> Lets say you wanted a developer experience similar to solidity, but uses WASM instead. What would be required on the VM side to support this?

A `hypervm` would need to provide a way to invoke a WASM runtime (like wasmer) from one of their `Actions`. They would also need to provide a way to deploy a WASM blob (the program that would be executed by the WASM runtime) and implement some sort of storage backend that persists these blobs and the state of the blob (like token balances of a smart contract).

The hope is for the `hypersdk` to provide all of these things `out-of-the-box` so that any `hypervm` can just plug-and-play programmability into their own blockchain. We plan to kick off this work just after we merge Avalanche Warp Messaging (AWM) support. We're tracking this work here: https://github.com/ava-labs/hypersdk/issues/65

> Would a smart contract written in GO for example, be deployable on every hypervm (assuming wasm compatibilty is built in)?

If all `hypervms` end up using the shared module provided by the `hypersdk`, they would be compatible ❤️ .

(Sorry for the delay responding...still figuring out how to get notifications)

**Comment by patrick-ogrady** [comment]
> What does the process look like when you get higher up on the stack?

A lot of this is not well-defined yet. We are focusing on ensuring the core is reliable and fast before building out all the dev tooling/support on top (as primitives approaches may change significantly while we mature the codebase).

> What kind of tooling/SDKs would work with a hyperVM , what kind wouldn't?
So far, any Golang package could be embedded into a a `hypervm` as long as it is deterministic (if not deterministic, execution could cause a chain split). Also, any Golang-related debugging/testing tools work out-of-the-box (as this is just standard Golang code).

No EVM-related tooling works for the `hypersdk` because there are no EVM-related components.

I think the biggest missing "UX" piece rn is a simple `hypersdk-compatible` wallet that requires the use of some specific `Auth` modules.
